src/powerset.rs

```diff
@@ -0,0 +1,85 @@
+use std::fmt;
+use alloc::vec::Vec;
+use std::usize;
+
+use super::combinations::Combinations;
+use super::size_hint;
+
+/// An iterator over the powerset of all elements from its source iterator.
+///
+/// See [`.powerset()`](../trait.Itertools.html#method.powerset) for more
+/// information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct Powerset<I: Iterator> {
+    combs: Combinations<I>,
+    pos: usize,
+}
+
+impl<I> Clone for Powerset<I>
+    where I: Clone + Iterator,
+          I::Item: Clone,
+{
+    clone_fields!(combs, pos);
+}
+
+impl<I> fmt::Debug for Powerset<I>
+    where I: Iterator + fmt::Debug,
+          I::Item: fmt::Debug,
+{
+    debug_fmt_fields!(Powerset, combs, pos);
+}
+
+/// Create a new `Powerset` from a cloneable iterator.
+pub fn powerset<I>(src: I) -> Powerset<I>
+    where I: Iterator,
+          I::Item: Clone,
+{
+    Powerset {
+        combs: Combinations::new(src, 0),
+        pos: 0,
+    }
+}
+
+impl<I> Iterator for Powerset<I>
+    where
+        I: Iterator,
+        I::Item: Clone,
+{
+    type Item = Vec<I::Item>;
+
+    fn next(&mut self) -> Option<Vec<I::Item>> {
+        if let Some(elt) = self.combs.next() {
+            self.pos = self.pos.saturating_add(1);
+            Some(elt)
+        } else {
+            if self.combs.k() < self.combs.n() || self.combs.k() == 0 {
```

Comment:
Is the `self.combs.k()==0` check necessary? If `k==0`, then line 51 should yield the empty subset (`[]`), and the next step should increase `k` to see if we need to continue - in particular also if `k<n`, right?

Reply:
Yes it's necessary I think. `k == 0` is a special case (first element) where `k == n == 0` and so `k < n` is false. Without it we'd never try to increase `k`.

The alternative is to change the preceding part of the condition on line 55 to be `k <= n` which would have the effect of a pointless `Combinations` that has `k > n` at the end of a `Powerset`'s iteration.

Since it's the special case I've placed it last so that it get's short-circuited and split the condition across two lines for clarity.

Reply:
Ok. But now we have special case k>n only for n==k==0. I guess ideally we would just increment k unconditionally and let Combinations handle the case k>n - which it already does anyway.

The only caveat I could imagine with this simple approach are overflow related issues. Maybe @jswrenn can steer you in the right direction.

Reply:
The thing stopping me from taking the approach of incrementing `k` unconditionally (i.e. pushing to `indices`) is that it would cause `indices` to allocate/copy unnecessarily when it's at capacity.

Reply:
Let's avoid the needless allocation.